docker, containerd: image registry mirrors (#1629)
bottlerocket-os/bottlerocket: README.md

  Supported values are `debug`, `info`, `warn`, `error`, and `crit`, and the default is `info`.
* `settings.ecs.enable-spot-instance-draining`: If the instance receives a spot termination notice, the agent will set the instance's state to `DRAINING`, so the workload can be moved gracefully before the instance is removed. Defaults to `false`.

#### Image registry settings

The following setting is optional and allows you to configure image registry mirrors and pull-through caches for your orchestrated containers.
* `settings.image-registry.mirrors`: A mapping of image registry to a list of image registry URL endpoints.  When pulling an image from a registry, the container runtime will try the endpoints one by one and use the first working one.
  * Example user data for setting up image registry mirrors:
  ```
  [settings.image-registry.mirrors]
  "docker.io" = ["https://<my-docker-hub-mirror-host>", "https://registry-1.docker.io"]

>   "gcr.io" = ["https://<example-gcr-mirror>","https://<example-gcr-mirror-2>"]

Reviewer: ```suggestion
  "gcr.io" = ["https://<my-gcr-mirror-host>", "https://<example-gcr-mirror-2>"]
```
The suggestion is a nit, but to make the examples consistent so no one thinks there's an important difference.

Question, though - could we use `https://gcr.io` as the second entry in the gcr list, so it's a working entry, like the dockerhub URL in the line above?

Author: > Question, though - could we use https://gcr.io as the second entry in the gcr list, so it's a working entry, like the dockerhub URL in the line above?

Yes, we can. But it's also redundant (at least for containerd) since it would try the registry directly if all the mirrors fail. For docker it's not clear so I listed docker hub in the list as well as a fallback.

Reviewer: Ah, ok, that's interesting.  If it's quick to confirm the difference, that'd be nice.  Maybe we could add a comment in the README above this, if it's unexpected?

Author: I just tried setting the mirror to some gibberish and ECS was still able to run a task on the node so I guess it fallsback by default. I'll go ahead and fix the example.

Reviewer: It might be worth clarifying (with a comment in README) that it will still fall back to official sources - this isn't a way to insulate yourself from third parties.  That could be useful from the perspectives of debugging and security.

  ```

>   Note that for variants that use Docker as its container runtime, Docker only supports pull-through caches for images from Docker Hub (docker.io). Registry mirrors for other registries are ignored in this case.

Reviewer: ```suggestion
  If you use a Bottlerocket variant that uses Docker as the container runtime, like aws-ecs-1, you should be aware that Docker only supports pull-through caches for images from Docker Hub (docker.io).  Mirrors for other registries are ignored in this case.
```

  Supported values are `debug`, `info`, `warn`, `error`, and `crit`, and the default is `info`.
* `settings.ecs.enable-spot-instance-draining`: If the instance receives a spot termination notice, the agent will set the instance's state to `DRAINING`, so the workload can be moved gracefully before the instance is removed. Defaults to `false`.

> #### Image registry settings

Reviewer 2: We talk about OS images in this readme too, so it might be worthwhile to call out that this is specific to container images.

```suggestion
#### Container image registry settings
```

> The following setting is optional and allows you to configure image registry mirrors and pull-through caches for your orchestrated containers.

Reviewer 2: Does this only cover orchestrated containers and not host containers?  Should it cover both?

Author: Right, this currently only covers orchestrated containers (containers via cri-containerd). 

For host-containers it's not immediately clear to me how I can configure the containerd client to use registry mirrors. I suspect I'd have to modify resolver to accomplish that? I can create a separate issue to track registry mirrors for host-containers.

Reviewer 2: Yes, this would be something that would have to be done in `host-ctr`.  I'm not sure if we should have the setting called `settings.container-registry.mirrors` if it only controls orchestrated containers though; how would we change this to handle host containers in the future?

Reviewer 3: It seems like it's just a bug that `host-ctr` and host containers wouldn't support the mirror settings - same as if a component doesn't support the system-wide proxy settings.

* `settings.image-registry.mirrors`: A mapping of image registry to a list of image registry URL endpoints.  When pulling an image from a registry, the container runtime will try the endpoints one by one and use the first working one.

>   Even if the default registry is not specified in the list, docker and containerd will still automatically try it at the end with the `https` scheme.

Reviewer: ```suggestion
  (Docker and containerd will still try the default registry URL if the mirrors fail.)
```
I was a bit confused about the meaning; I think this is clearer?

  * Example user data for setting up image registry mirrors:
  ```
  [settings.image-registry.mirrors]
  "docker.io" = ["https://<my-docker-hub-mirror-host>"]
  "gcr.io" = ["https://<my-gcr-mirror-host>","http://<my-gcr-mirror-host-2>"]

>   ```

Reviewer 3: 🚲 🏠 ⚠️

I saw you fixed this in the README , but `container-registry` might be a better term for the setting since "images" are kind of ambiguous.

In the context of a running Bottlerocket it's probably clear enough, but we deal with a lot of different image types - full disk images, filesystem images, etc.
